### ISSUE TITLE
Overriding first playlist selection

### ISSUE THREAD

**Issue by vasklund** [videojs/videojs-contrib-hls#277]
We are having problems with the initial segment almost always being downloaded in the lowest bitrate, even when available bandwidth allows for playback at higher quality. This is a pretty big problem for short clips (a lot of our clips are around 30-120 sec long).

I know we can override `selectPlaylist` to implement our own selection algorithm, or set `player.hls.bandwidth` higher to make `selectPlaylist` return a higher quality playlist. The problem is that the code in `this.playlists.on('loadedmetadata', function { ... })`, which is responsible for selecting the first playlist, will not respect the playlist returned from `selectPlaylist` if the estimated download time for the first segment is more than 10 seconds. This seems reasonable, but the problem is that it will then fallback to the **lowest** bitrate, instead of selecting the highest quality playlist that satisfies both available bandwidth and initial load time (10 seconds).

Essentially, we want to trade high quality (from the first segment) for longer startup time, but there is no good way to do so now. I'm experimenting with some alternate approaches and will try to prepare a pull request today, but I'm interested in getting some feedback.

**Comment by vasklund** [comment]
https://github.com/videojs/videojs-contrib-hls/blob/master/src/videojs-hls.js#L193

**Comment by vasklund** [comment]
Ok, It seems I'm able to sidestep the problem by setting the playlist manually with `player.hls.playlists.media`:

```
player.ready(function () {
  // force the fourth playlist for the initial segment
  player.hls.playlists.media(player.hls.playlists.master.playlists[3]);
});
```

This seems to work and I can leave player.hls.selectPlaylist and player.hls.bandwidth alone, but it seems kind of counterintuitive that I need to use `player.hls.playlists.media()` to set the quality for the first segment. I thought I would be able to achieve similar results by overriding `player.hls.selectPlaylist()`, but it seems I can't override it before it gets called (if I override it in player.ready() it has already been called once for the initial segment).

I'll comment again if I find anything else, but for now I might just go with the media() workaround.

**Comment by dmlap** [comment]
Did you notice that you can influence initial rendition selection by [overriding the bandwidth estimate](https://github.com/videojs/videojs-contrib-hls#playerhlsbandwidth) on `loadstart`?

```js
player.on('loadstart', function() {
  // choose a really high quality rendition, even if bandwidth estimates don't justify it
  player.hls.bandwidth = 1e9;
});
```
If your media() solution or the bandwidth-based one don't solve your problem, please comment in here. This is something we would like to ensure is easy to do.

**Comment by vasklund** [comment]
I'd like to reopen this, since I'm still not able to override the initial quality selection.

The problem in both cases (using `player.hls.playlists.media(myHighQualityPlaylist)` and the bandwidth spoofing above) is that there's no good event to hook the custom logic to.

The bandwidth spoof seems like the easiest implementation, but the problem is that the `loadstarted` event sometimes happens __after__ the `loadedmetadata` event. The bandwidth spoof has no effect in this case since it is read in the (playlists) `loadedmetadata` handler ([videojs-hls.js#L157](https://github.com/videojs/videojs-contrib-hls/blob/c18e3df75d44d5552ce0521f37bbbfd8bbfc6c5e/src/videojs-hls.js#L157)), where it is still `undefined`.

I set up a simple test page on:
http://jsbin.com/xedeyovecu/1/
I see the error about 50% of the time (loadedmetadata happens after loadstart), on a couple of different browser/OS combinations, so hopefully you should be able to reproduce easily with a couple of refreshes.

**Comment by vasklund** [comment]
I might be able to spend some time on this, but before that I'd like to ask whether the behavior above (`loadedmetadata` sometimes comes before `loadstarted`) is expected behavior or not? If not I could look into it; if it is I'm not really sure how we want to approach the problem described in the issue.

**Comment by dmlap** [comment]
`loadedmetadata` should definitely not be coming before `loadstart`. If you're able to dig into this a bit deeper, we'd definitely appreciate it. Thanks!

**Comment by vasklund** [comment]
Looking at this now and have a quick follow-up question:
I'm seeing the `loadedmetadata` event twice, once from [videojs-hls.js#L146](https://github.com/videojs/videojs-contrib-hls/blob/c18e3df75d44d5552ce0521f37bbbfd8bbfc6c5e/src/videojs-hls.js#L146) and once from Flash via `vjs.Flash['onEvent']`. I was expecting the event only once, but I haven't read the specification that closely. Do you know if this is intended or not? 

I've just started digging into this so I don't know if it's intended or if it has any negative implications.

**Comment by dmlap** [comment]
Ah, good find! You're right, `loadedmetadata` should only be filed once per video load; that is not intended behavior.

**Comment by vasklund** [comment]
I haven't been able to find a good solution to this yet, but I've documented what I think is the root problem in a new issue at #310. We've sidestepped the problem by monkey-patching videojs.Hls.prototype.src for now (not a wonderful solution), so this issue can be closed at your discretion.

**Comment by dmlap** [comment]
Should be fixed via #356
